[Release] Add daily history models and history functionality to Salesforce transform package

fivetran-avinash opened the pull request:
## PR Overview
**This PR will address the following Issue/Feature:** This is updaets to include Salesforce History Mode in the existing Salesforce packages. 

**This PR will result in the following new package version:** 0.10.0
<!--- Please add details around your decision for breaking vs non-breaking version upgrade. If this is a breaking change, were backwards-compatible options explored? -->

**Please detail what change(s) this PR introduces and any additional information that should be known during the review of this PR:** Add account, contact, opportunity daily history models along with ability to switch between history and regular Salesforce models.

## PR Checklist
### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt compile
- [x] dbt run –full-refresh
- [x] dbt run
- [x] dbt test
- [x] dbt run –vars (if applicable)

Before marking this PR as "ready for review" the following have been applied:
- [ ] The appropriate issue has been linked and tagged  (no existing issue, should i create one?)
- [x] You are assigned to the corresponding issue and this PR
- [x] BuildKite integration tests are passing

### Detailed Validation
Please acknowledge that the following validation checks have been performed prior to marking this PR as "ready for review":
- [x] You have validated these changes and assure this PR will address the respective Issue/Feature.
- [x] You are reasonably confident these changes will not impact any other components of this package or any dependent packages.
- [x] You have provided details below around the validation steps performed to gain confidence in these changes.
<!--- Provide the steps you took to validate your changes below. -->
1) Run models as if no changes applied to ensure config works
2) Change schema configuration to run history models
3) Run specific history models
4) Apply passthrough columns
5) Apply filters to make sure date filter function on temp models work

### Standard Updates
Please acknowledge that your PR contains the following standard updates:
- Package versioning has been appropriately indexed in the following locations:
    - [x] indexed within dbt_project.yml
    - [x] indexed within integration_tests/dbt_project.yml
- [x] CHANGELOG has individual entries for each respective change in this PR
    <!--- If there is a parallel upstream change, remember to reference the corresponding CHANGELOG as an individual entry.  -->
- [x] README updates have been applied (if applicable)
    <!--- Remember to check the following README locations for common updates. →
        <!--- Suggested install range (needed for breaking changes) →
        <!--- Dependency matrix is appropriately updated (if applicable) →
        <!--- New variable documentation (if applicable) -->
- [NA] DECISIONLOG updates have been updated (if applicable)
- [x] Appropriate yml documentation has been added (if applicable)

### dbt Docs
Please acknowledge that after the above were all completed the below were applied to your branch:
- [x] docs were regenerated (unless this PR does not include any code or yml updates)

### If you had to summarize this PR in an emoji, which would it be?
<!--- For a complete list of markdown compatible emojis check our this git repo (https://gist.github.com/rxaviers/7360908)  --> 
🍾

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-avinash thanks so much for working through this! I have a couple comments below that I would like you to review and respond to. There are a few comments that will require us to sync in more detail on. Let's discuss in person on Monday.

fivetran-avinash commented on CHANGELOG.md:
```diff
@@ -3,6 +3,26 @@ Please note that this data model will now create a new transformation for **all*
+
+- Customers now can configure their Salesforce package to pull from their Salesforce History Mode source tables (as opposed to the default Salesforce source tables) and populate their existing models in the packages. More instructions can be found in the README. 
+
+- These models are disabled by default, so you will have to enable the equivalent models below in your `dbt_project.yml` to utilize them.  
```
Yes, forgot to include the yml, added in now.

fivetran-avinash commented on README.md:
```diff
@@ -16,30 +16,36 @@
-| [salesforce__opportunity_line_item_enhanced](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__opportunity_line_item_enhanced)  |Each record represents a line item belonging to a certain opportunity, with additional product details.
-
-**Note**: For Quickstart Data Model users only, in addition to the above output models you will also receive models in your transformation list which replicate **all** of your Salesforce objects with the inclusion of the relevant formula fields in the generated output models.
+|**model**|**description**|**Quickstart compatible**
+-----|-----|-----
+| [salesforce__manager_performance](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__manager_performance)     |Each record represents a manager, enriched with data about their team's pipeline, bookings, losses, and win percentages. | Yes
+| [salesforce__owner_performance](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__owner_performance)         |Each record represents an individual member of the sales team, enriched with data about their pipeline, bookings, losses, and win percentages. | Yes
+| [salesforce__sales_snapshot](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__sales_snapshot)               |A single row snapshot that provides various metrics about your sales funnel. | Yes
+| [salesforce__opportunity_enhanced](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__opportunity_enhanced)  |Each record represents an opportunity, enriched with related data about the account and opportunity owner. | Yes
+| [salesforce__contact_enhanced](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__contact_enhanced)  |Each record represents a contact with additional account and owner information. | Yes
+| [salesforce__daily_activity](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__daily_activity)  |Each record represents a daily summary of the number of sales activities, for example tasks and opportunities closed. | Yes
+| [salesforce__opportunity_line_item_enhanced](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__opportunity_line_item_enhanced)  |Each record represents a line item belonging to a certain opportunity, with additional product details. | Yes
+| [salesforce__account_daily_history](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__account_daily_history) | Each record is a daily record in an account, starting with its first active date and updating up toward either the current date (if still active) or its last active date. | No
+| [salesforce__contact_daily_history](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__contact_daily_history) |  Each record is a daily record in an contact, starting with its first active date and updating up toward either the current date (if still active) or its last active date.
+ | No
+| [salesforce__opportunity_daily_history](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__opportunity_daily_history) | Each record is a daily record in an opportunity, starting with its first active date and updating up toward either the current date (if still active) or its last active date. | No
+ 
+**Note**: For Quickstart Data Model users only, in addition to the above output models that are Quickstart compatible, you will also receive models in your transformation list which replicate **all** of your Salesforce objects with the inclusion of the relevant formula fields in the generated output models.
```
Fixed.

fivetran-avinash commented on CHANGELOG.md:
```diff
@@ -3,6 +3,26 @@ Please note that this data model will now create a new transformation for **all*
 
+# dbt_salesforce v0.10.0 
+## 🚨 Breaking Changes 🚨:
+- We have added daily history mode models in the [`models/history`](https://github.com/fivetran/dbt_salesforce_source/tree/main/models/history) folder [based off of Fivetran's history mode feature](https://fivetran.com/docs/core-concepts/sync-modes/history-mode), pulling from source models in `dbt_salesforce_source`. This will allow customers to utilize the Fivetran history mode feature, which re ...
```
Updated.

fivetran-avinash commented on README.md:
```diff
@@ -16,30 +16,36 @@
-| [salesforce__opportunity_line_item_enhanced](https://fivetran.github.io/dbt_salesforce/#!/model/model.salesforce.salesforce__opportunity_line_item_enhanced)  |Each record represents a line item belonging to a certain opportunity, with additional product details.
-
-**Note**: For Quickstart Data Model users only, in addition to the above output models you will also receive models in your transformation list which replicate **all** of your Salesforce objects with the inclusion of the relevant formula fields in the generated output models.
+|**model**|**description**|**Quickstart compatible**
```
sounds good to me!

fivetran-avinash commented on README.md:
````diff
@@ -74,6 +80,29 @@ vars:
+
+If you wish to switch to utilizing ONLY Salesforce History Mode source tables, you should change the default config set up in your source yml files within your `dbt_project.yml`.
+
+```yml
+sources:
+  salesforce_source: 
+    salesforce:
+      +enabled: false # True by default. Disable if you wish to not utilize the default Salesforce source tables.
+    salesforce_history:
+      +enabled: true # False by default. Enable if you wish to utilize the default Salesforce source tables.
+``` 
````
Removed.

fivetran-avinash commented on models/intermediate/int_salesforce__date_spine.sql:
```diff
@@ -3,30 +3,35 @@ with spine as (
-        {% if var('salesforce__lead_enabled', True) %}
+        {% if var('not_using_salesforce_history_mode', True) and var('salesforce__lead_enabled', True) %}
             select  min( created_date ) as min_date from {{ source('salesforce', 'lead') }}
-
-        {% else %}
+        {% elif var('not_using_salesforce_history_mode', True) and var('salesforce__lead_enabled', False) %}
             select  min( created_date ) as min_date from {{ source('salesforce', 'opportunity') }}
+        {% elif var('not_using_salesforce_history_mode', False) and var('salesforce__lead_history_enabled', True) %}
+            select  min( created_date ) as min_date from {{ source('salesforce_history', 'lead') }}
+        {% else %}
+            select  min( created_date ) as min_date from {{ source('salesforce_history', 'opportunity') }}
```
Agreed, these are now removed.

fivetran-avinash commented on models/history/salesforce__contact_daily_history.sql:
```diff
@@ -0,0 +1,87 @@
+    {% if not var('contact_history_first_date', None) or not var('contact_history_last_date', None) %}
+        {% set date_query %}
+        select 
+            min( _fivetran_start ) as min_date,
```
@fivetran-joemarkiewicz 
Hmm, so a default date to start? Should be doable. 

Should we have a default start date from the inception of Salesforce (March 1997), Fivetran (2013?), or something customized ([Jira has 2016](https://github.com/fivetran/dbt_jira/blob/main/models/intermediate/field_history/int_jira__issue_calendar_spine.sql#L21C16-L21C51))?

Something like 
```

        {% set date_query %}
        select 
        {% if  _fivetran_start > DEFAULT DATE %}
             min( _fivetran_start ) as min_date
         {% else % } 
             DEFAULT_DATE as min_date
        {% endif %} 
        {{ dbt.date_trunc('day', dbt.current_timestamp_backcompat()) }} as max_date
        from {{ source('salesforce_history', 'account') }}
        {% endset %}
```

fivetran-joemarkiewicz replied:
Let's just be consistent with Jira and have the default start date be 2016. It makes sense to have it start at the beginning of Salesforce creation; however, that is really far back and could possibly create a huge date spine when people don't actually care about that. 2016 seems like a good middle ground of a lot of history, but not too far back.

fivetran-avinash commented on models/intermediate/int_salesforce__opportunity_aggregation_by_owner.sql:
```diff
@@ -7,7 +7,12 @@ with salesforce_opportunity_enhanced as (
 salesforce_user as (
 
     select *
+    {% if var('not_using_salesforce_history_mode', True) %}
     from {{ var('user') }}
+    {% else %}
+    from {{ var('user_history') }}
+    where _fivetran_active = true
+    {% endif %}
```
Agreed.

fivetran-avinash commented on models/salesforce__contact_enhanced.sql:
```diff
@@ -1,21 +1,55 @@
     select *
+    {% if var('not_using_salesforce_history_mode', True) %}
     from {{ var('user') }}
+    {% else %}
+    from {{ var('user_history') }}
+    where _fivetran_active = true
+    {% endif %}
+),
+
+contact_seed as (
+
+    select *
+    from {{ var('contact') }}
+),
+
+account_seed as (
+
+    select *
+    from {{ var('account') }}
+),
+
+user_seed as (
+
+    select *
+    from {{ var('account') }}
```
Localized testing, removed now.

fivetran-avinash commented on models/history/salesforce__contact_daily_history.sql:
```diff
@@ -0,0 +1,87 @@
+with spine as (
+
+    {% if execute %}
+    {% if not var('contact_history_first_date', None) or not var('contact_history_last_date', None) %}
```
Resolved in person, but for context sake: The earliest `_fivetran_start` date should have historical records preceding it. The earliest `created_at` date doesn't necessarily have historical records, as history mode might not have been available by that point. That would lead to numerous null records.

fivetran-avinash commented on models/salesforce__daily_activity.sql:
```diff
@@ -5,35 +5,73 @@ with date_spine as (
 ),
 
-{% if var('salesforce__task_enabled', True) %}
+task_seed as (
+    select *
+    from {{ var('task') }}
+),
+
+
+event_seed as (
+    select *
+    from {{ var('event') }}
+),
+
+lead_seed as (
+    select *
+    from {{ var('lead') }}
+),
+
+opportunity_seed as (
+    select *
+    from {{ var('opportunity') }}
+),
```
Same comment as above, these are removed.

fivetran-avinash commented on models/salesforce__contact_enhanced.sql:
```diff
@@ -1,21 +1,55 @@
 with contact as (
 
     select *
+    {% if var('not_using_salesforce_history_mode', True) %}
     from {{ var('contact') }}
+    {% else %}
+    from {{ var('contact_history') }}
+    where _fivetran_active = true
+    {% endif %}
```
Wiped the slate clean!

fivetran-avinash commented on models/history/salesforce__contact_daily_history.sql:
```diff
@@ -0,0 +1,87 @@
+        and get_latest_daily_value._fivetran_end >= cast(spine.date_day as {{ dbt.type_timestamp() }})
+)
+
+select * 
+from daily_history
```
Agreed, and this has been implemented in the source package PR!

fivetran-avinash commented on models/history/salesforce__contact_daily_history.sql:
```diff
@@ -0,0 +1,87 @@
+{{ config(enabled=var('salesforce__contact_history_enabled', False)) }}
+
+{{
+    config(
+        materialized='incremental',
+        partition_by = {'field': 'date_day', 'data_type': 'date'}
+            if target.type not in ['spark', 'databricks'] else ['date_day'],
+        unique_key='contact_day_id',
+        incremental_strategy = 'merge' if target.type not in ('snowflake', 'postgres', 'redshift') else 'delete+insert',
+        file_format = 'delta'
+    )
+}}
```
Updated on all models.

fivetran-avinash commented on README.md:
````diff
@@ -101,8 +130,27 @@ Source tables are referenced using default names. If an individual source table
+  salesforce__user_role_history_enabled: true        # False by default. Only use if you have history mode enabled and wish to leverage the user role history table.
+```
+
+### Salesforce History Mode Active Records
+If you have Salesforce [History Mode](https://fivetran.com/docs/getting-started/feature/history-mode) enabled for your connector, in your `dbt_project.yml`, you will want to add and set the desired `using_[table]_history_mode_active_records` variable(s) as `true` to filter for only active records as the standard salesforce models are designed for non-historical data. These variables are disabled  ...
````
This functionality has been deprecated, so excising this section.

fivetran-avinash commented on README.md:
````diff
@@ -101,8 +130,27 @@ Source tables are referenced using default names. If an individual source table
 ```
-### Salesforce History Mode
-If you have Salesforce [History Mode](https://fivetran.com/docs/getting-started/feature/history-mode) enabled for your connector, in your `dbt_project.yml`, you will want to add and set the desired `using_[table]_history_mode_active_records` variable(s) as `true` to filter for only active records as the package is designed for non-historical data. These variables are disabled by default. 
+### Salesforce History Mode Daily Models
+If you have Salesforce [History Mode](https://fivetran.com/docs/getting-started/feature/history-mode) enabled for your connector, in your `dbt_project.yml`, you will want to add the following variables as `true` to enable these models. These variables are disabled by default. In particular, `salesforce__account_history_enabled`, `salesforce__contact_history_enabled` and `salesforce__opportunity_h ...
+
+```yml
+# dbt_project.yml
+
+...
+vars:
+  salesforce__account_history_enabled: true      # False by default. Only use if you have history mode enabled and wish to leverage the account history table.
+  salesforce__contact_history_enabled: true  # False by default. Only use if you have history mode enabled and wish to leverage the contact history table.
+  salesforce__event_history_enabled: true    # False by default. Only use if you have history mode enabled and wish to leverage the event history table.
+  salesforce__lead_history_enabled: true         # False by default. Only use if you have history mode enabled and wish to leverage the lead history table.
+  salesforce__opportunity_history_enabled: true      # False by default. Only use if you have history mode enabled and wish to leverage the opportunity history table.
+  salesforce__task_history_enabled: true         # False by default. Only use if you have history mode enabled and wish to leverage the task history table.
+  salesforce__user_history_enabled: true         # False by default. Only use if you have history mode enabled and wish to leverage the user history table.
+  salesforce__user_role_history_enabled: true        # False by default. Only use if you have history mode enabled and wish to leverage the user role history table.
+```
````
Consolidated. Also removing variables no longer being used.

fivetran-avinash commented on README.md:
````diff
@@ -74,6 +80,29 @@ vars:
+    salesforce_history_schema: your_history_schema_name
+```
+
+If you wish to switch to utilizing ONLY Salesforce History Mode source tables, you should change the default config set up in your source yml files within your `dbt_project.yml`.
````
Updated this with the new README section created in the source section, with slight modifications.

fivetran-avinash commented on models/history/salesforce__contact_daily_history.sql:
```diff
@@ -0,0 +1,87 @@
+    {% endif %}
+),
+
+contact_history as (
+
+    select *,
+        cast( {{ dbt.date_trunc('day', '_fivetran_start') }} as date) as start_day       
+    from {{ var('contact_history') }}
```
D'oh. Yes. Updated.

fivetran-avinash commented:
@fivetran-joemarkiewicz I have addressed all transform notes! There is one open question about new date spine logic ([see my comment here](https://github.com/fivetran/dbt_salesforce/pull/45#discussion_r1372480467)) but we can try and resolve today.

One additional update: I moved up the IMPORTANT section of the Readme to near the start of the History mode section in each package. I feel customers leveraging history mode should know this upfront and we should highlight it early before attempting any configuration.

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -3,6 +3,39 @@ Please note that this data model will now create a new transformation for **all*
 
+# dbt_salesforce v1.0.0 
+
+**📣 THIS IS A MAJOR PACKAGE RELEASE! 📣** More details below. 
```
Please see my source package notes for similar CHANGELOG updates that should also be applied here.

fivetran-avinash replied:
Updated.

fivetran-joemarkiewicz commented on CHANGELOG.md:
```diff
@@ -3,6 +3,39 @@ Please note that this data model will now create a new transformation for **all*
+# dbt_salesforce v1.0.0 
+
+**📣 THIS IS A MAJOR PACKAGE RELEASE! 📣** More details below. 
+## 🚨 Breaking Changes 🚨:
```
We will also want to highlight the breaking change in the removal of the tmp models from the source here.

fivetran-avinash replied:
Updated

fivetran-joemarkiewicz commented on README.md:
````diff
@@ -16,51 +16,66 @@
 ```
+
+### Database Incremental Strategies 
+Some of the end models in this package are materialized incrementally. We have chosen `insert_overwrite` as the default strategy for **BigQuery** and **Databricks** databases, as it is only available for these dbt adapters. For **Snowflake**, **Redshift**, and **Postgres** databases, we have chosen `delete+insert` as the default strategy.
````
```suggestion
The history end models in this package are materialized incrementally. We have chosen `insert_overwrite` as the default strategy for **BigQuery** and **Databricks** databases, as it is only available for these dbt adapters. For **Snowflake**, **Redshift**, and **Postgres** databases, we have chosen `delete+insert` as the default strategy.
```

fivetran-avinash replied:
Committed

fivetran-joemarkiewicz commented on README.md:
````diff
@@ -92,6 +107,84 @@ vars:
 ```
 The corresponding metrics from the disabled tables will not populate in the downstream models.
 
+## Optional: Utilizing Salesforce History Mode records
````
Please see my source package review notes for similar updates that should be applied to this README as well.

fivetran-avinash replied:
Updated

fivetran-joemarkiewicz commented on dbt_project.yml:
```diff
@@ -21,6 +22,17 @@ vars:
+    event_history: "{{ ref('stg_salesforce__event_history') }}"
+    lead_history: "{{ ref('stg_salesforce__lead_history') }}"
+    opportunity_history: "{{ ref('stg_salesforce__opportunity_history') }}"
+    opportunity_line_item_history: "{{ ref('stg_salesforce__opportunity_line_item_history') }}"
```
Same comment about removing history variables that are not used in the end

fivetran-avinash replied:
Updated

fivetran-joemarkiewicz commented on integration_tests/dbt_project.yml:
```diff
@@ -23,7 +30,18 @@ vars:
+    ##history mode identifiers
+    salesforce_account_history_identifier: "sf_account_history_data"
+    salesforce_contact_history_identifier: "sf_contact_history_data"
+    salesforce_event_history_identifier: "sf_event_history_data"
```
Same comment about removing seeds and identifiers for history models that are no longer used.

fivetran-joemarkiewicz commented on integration_tests/dbt_project.yml:
```diff
@@ -145,6 +163,81 @@ seeds:
+        last_modified_date: timestamp 
+        last_viewed_date: timestamp
+        is_deleted: boolean
+    sf_event_history_data:
```
Same comment about removing the history configs for models we are not bringing forward anymore.

fivetran-avinash replied:
Removed

fivetran-joemarkiewicz commented on integration_tests/seeds/sf_order_history_data.csv:
```diff
@@ -0,0 +1,11 @@
+_fivetran_active,_fivetran_start,_fivetran_end,id,_fivetran_synced,account_id,activated_by_id,activated_date,billing_city,billing_country,billing_country_code,billing_geocode_accuracy,billing_latitude,billing_longitude,billing_postal_code,billing_state,billing_state_code,billing_street,company_authorized_by_id,contract_id,created_by_id,created_date,customer_authorized_by_id,description,effective_ ...
```
Let's remove this since we are not using it in the end models anymore.

fivetran-joemarkiewicz commented on integration_tests/seeds/sf_event_history_data.csv:
```diff
@@ -0,0 +1,11 @@
+_fivetran_active,_fivetran_start,_fivetran_end,id,_fivetran_synced,account_id,activity_date,activity_date_time,created_by_id,created_date,description,end_date_time,event_subtype,group_event_type,invitee_uuid_c,is_child,is_deleted,is_group_event,is_recurrence,is_reminder_set,last_modified_by_id,last_modified_date,location,no_show_c,owner_id,start_date_time,subject,type,what_count,what_id,who_count ...
```
Let's remove this since we are not using it in the end models anymore.

fivetran-avinash replied:
Removed

fivetran-joemarkiewicz commented on integration_tests/seeds/sf_lead_history_data.csv:
```diff
@@ -0,0 +1,2 @@
+_fivetran_active,_fivetran_start,_fivetran_end,country,email_bounced_reason,email_bounced_date,owner_id,lead_source,converted_date,last_modified_date,master_record_id,last_modified_by_id,geocode_accuracy,converted_contact_id,id,state,last_referenced_date,last_activity_date,phone,name,created_by_id,is_converted,is_unread_by_owner,status,city,latitude,industry,title,_fivetran_synced,last_viewed_dat ...
```
Let's remove this since we are not using it in the end models anymore.

fivetran-avinash replied:
Removed

fivetran-joemarkiewicz commented on integration_tests/seeds/sf_opportunity_line_item_history_data.csv:
```diff
@@ -0,0 +1,11 @@
+_fivetran_active,_fivetran_start,_fivetran_end,product_2_id,quantity,total_price,description,discount,last_modified_date,list_price,product_code,unit_price,_fivetran_synced,is_deleted,opportunity_id,last_modified_by_id,service_date,system_modstamp,name,id,created_by_id,created_date,sort_order,pricebook_entry_id,one_saas_app_included_c,one_saas_app_quantity_invoiced_c,one_saas_app_quantity_not_inv ...
```
Let's remove this since we are not using it in the end models anymore.

fivetran-joemarkiewicz commented on integration_tests/seeds/sf_product_2_history_data.csv:
```diff
@@ -0,0 +1,3 @@
+_fivetran_active,_fivetran_start,_fivetran_end,last_referenced_date,is_active,description,last_modified_date,record_type_id,product_code,_fivetran_synced,last_viewed_date,is_deleted,last_modified_by_id,system_modstamp,name,id,created_by_id,created_date,family,max_volume_c,min_volume_c,request_name_c,default_quantity_c,account_c,related_product_c,picklist_value_c,as_input_c,as_output_c,status_c,nu ...
```
Let's remove this since we are not using it in the end models anymore.

fivetran-avinash replied:
Removed

fivetran-joemarkiewicz commented on integration_tests/seeds/sf_task_history_data.csv:
```diff
@@ -0,0 +1,11 @@
+_fivetran_active,_fivetran_start,_fivetran_end,who_id,call_disposition,owner_id,subject,description,last_modified_date,what_count,_fivetran_synced,call_object,is_deleted,last_modified_by_id,system_modstamp,id,call_type,is_high_priority,is_closed,activity_date,priority,who_count,account_id,call_duration_in_seconds,created_by_id,created_date,what_id,task_subtype,status,type,record_type_id,completed ...
```
Let's remove this since we are not using it in the end models anymore.

fivetran-avinash replied:
Removed

fivetran-joemarkiewicz commented on integration_tests/seeds/sf_user_history_data.csv:
```diff
@@ -0,0 +1,11 @@
+_fivetran_active,_fivetran_start,_fivetran_end,_fivetran_deleted,_fivetran_synced,account_id,alias,city,company_name,contact_id,country,department,email,first_name,id,individual_id,is_active,last_login_date,last_name,last_referenced_date,last_viewed_date,manager_id,name,postal_code,profile_id,state,street,title,user_role_id,user_type,username
```
Let's remove this since we are not using it in the end models anymore.

fivetran-joemarkiewicz commented on integration_tests/seeds/sf_user_role_history_data.csv:
```diff
@@ -0,0 +1,11 @@
+_fivetran_active,_fivetran_start,_fivetran_end,_fivetran_deleted,_fivetran_synced,developer_name,forecast_user_id,id,name,opportunity_access_for_account_owner,parent_role_id,rollup_description
```
Let's remove this since we are not using it in the end models anymore.

fivetran-joemarkiewicz commented on packages.yml:
```diff
@@ -1,3 +1,7 @@
 packages:
-  - package: fivetran/salesforce_source
-    version: [">=0.7.0", "<0.8.0"]
+  # - package: fivetran/salesforce_source
+  #   version: [">=1.0.0", "<1.1.0"]
+
+  - git: https://github.com/fivetran/dbt_salesforce_source.git
+    revision: release/add-history-mode
+    warn-unpinned: false
```
Reminder to swap before release

fivetran-joemarkiewicz commented on models/salesforce_history/salesforce__account_daily_history.sql:
```diff
@@ -0,0 +1,100 @@
+with spine as (
+
+    {% if execute %}
+    {% if not var('account_history_start_date', None) or not var('account_history_end_date', None) %}
```
Should we also add a config here for the `global_history_start_date`?

fivetran-joemarkiewicz replied:
Additionally, should we include a `global_history_end_date` variable?

fivetran-avinash replied:
Config added, but it's not working in the way I'd like.

fivetran-joemarkiewicz commented on models/salesforce_history/salesforce__account_daily_history.sql:
```diff
@@ -0,0 +1,100 @@
+    {% if not var('account_history_start_date', None) or not var('account_history_end_date', None) %}
+        {% set date_query %}
+        select 
+            min( _fivetran_start ) as min_date,
```
This is where we should add the default date. If we want, we can also do a `greatest` calculation. So if the minimum `_fivetran_start` date is higher than 2016, then we pick that.

fivetran-avinash replied:
Added this additional logic in.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-avinash thanks for working through these! This is looking great! I have a few comments that you can see below. Some comments from the source should also be applied here. I also made comments in the account daily history. Any changes there should also be applied to the other history models.

fivetran-avinash commented:
@fivetran-joemarkiewicz Changes applied. I was able to fix the spine so that it'll work with the greatest function!

Now I'm running into weird errors where it can't find account history 🤔 . Nothing looks out of the ordinary from my setup that contact/opportunity would run but account wouldn't. If you could take a look that'd be great.

fivetran-joemarkiewicz reviewed (changes requested):
@fivetran-avinash thanks for working so hard on these updates! I was able to review and made some adjustments that can be merged into your branch within PR #46.

Please review my comments in that PR and once you are comfortable you can merge them into your branch, regen the docs, and once buildkite passes we will be good to approve this and release!

I didn't approve this PR just yet because I want your thoughts on my PR linked above. There are some changes that I want your opinions on. Thanks again!! Let me know once this is good to approve!

fivetran-joemarkiewicz reviewed (approved):
lgtm
